Atmel SAM: Introduce SMC driver

pdgendt opened the pull request:
Add support for the ATSAM4SA16C variant and add a driver for the Atmel SMC/EBI module.

Tested on a custom board

pdgendt committed:
soc: arm: atmel_sam: sam4s: add support for sam4sa16c

Add soc support for sam4sa16c variant

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
drivers: memc: Introduce Atmel SAM SMC/EBI driver

Add a driver to support external memory connected to the SMC port
for Atmel SAM devices.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
dts: sam4s: Add SMC peripheral

Add pinctrl definitions for the SMC peripheral and add the module
to the sam4s soc.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

nandojve commented:
Hi @pdgendt ,

Thank you for this contribution. It seems a very clean implementation. I hope can review soon when merge window is open again.
I'm missing few things:

 * It is required a board in tree be configured with this driver to allow CI tests. Is it possible enable IS66WV51216DALL support at SAM4S, or even SAM4E? Please, let us know at least what is your use case.
 * It is necessary create a test. It doesn't need to be so elaborated but without it won't be possible build the driver at CI with a valid config. My suggestion is add as a sample at `samples/drivers/smc_<use case>` where _use case_ can be _sram_, _sdram_ etc.
 * It will be necessary add an entry as **smc** at board yaml file to enable twister tests. Be aware that **smc** will be the filter entry in the test case. https://github.com/zephyrproject-rtos/zephyr/blob/655b0d21274a5277d00091cf59c0d591ec0df5bb/boards/arm/sam4s_xplained/sam4s_xplained.yaml#L9-L16
 * I think SAM4S and SAM4E have same SMC IP, could you confirm? If yes, could you add SMC node for SAM4E too.
 * How users will mapped memory?

pdgendt committed:
dts: sam4e: Add SMC peripheral

Add pinctrl definitions for the SMC peripheral (8) and add the
module to the sam4e soc.

IDs and addresses are taken from the datasheet.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
Hi @nandojve, thank you for the response. I've update the PR, please check my responses below.

> It is required a board in tree be configured with this driver to allow CI tests. Is it possible enable IS66WV51216DALL support at SAM4S, or even SAM4E? Please, let us know at least what is your use case.

I've added the `IS66WV51216DALL` to the sam4s_xplained board, however I don't have access to this HW, can you/someone test the changes made? I've taken IDs/addresses/timings from datasheets and example projects.

> It is necessary create a test. It doesn't need to be so elaborated but without it won't be possible build the driver at CI with a valid config. My suggestion is add as a sample at `samples/drivers/smc_<use case>` where _use case_ can be _sram_, _sdram_ etc.

Is this still required if it is added to an existing board where other tests will probably already cover this?

> It will be necessary add an entry as **smc** at board yaml file to enable twister tests. Be aware that **smc** will be the filter entry in the test case. https://github.com/zephyrproject-rtos/zephyr/blob/655b0d21274a5277d00091cf59c0d591ec0df5bb/boards/arm/sam4s_xplained/sam4s_xplained.yaml#L9-L16

Added.

> I think SAM4S and SAM4E have same SMC IP, could you confirm? If yes, could you add SMC node for SAM4E too.

Added based on the SAM4E datasheet, unable to test due to lack of HW.

> How users will mapped memory?

I've updated the `yaml` file on how to add the memory to the dts.

nandojve commented on drivers/memc/memc_sam_smc.c:
```diff
@@ -0,0 +1,103 @@
+	const struct memc_smc_config *cfg = dev->config;
+	SmcCs_number *bank;
+
+	/* Enable SMC clock in PMC */
```
I think all comments in this file could be dropped.

nandojve reviewed (changes requested):
Hi @pdgendt ,

It will be necessary add a test at `tests/drivers/memc`. My suggestion is do something simple like write a pattern, read back and compare. This could be a generic test that may be used for any future `memc_sram` driver inclusive.

pdgendt commented:
> It will be necessary add a test at `tests/drivers/memc`. My suggestion is do something simple like write a pattern, read back and compare. This could be a generic test that may be used for any future `memc_sram` driver inclusive.

Hey @nandojve,

I was looking into this, and there's a test for `stm32_sdram` which could be reused. However this requires some more effort for the linker sections to be more generic.

Looking at #42008 and #40422 we could rework that existing test once those PRs are merged.

nandojve commented on boards/arm/sam4s_xplained/sam4s_xplained_defconfig:
```diff
@@ -12,6 +12,7 @@ CONFIG_BOARD_SAM4S_XPLAINED=y
 CONFIG_SOC_ATMEL_SAM4S_EXT_MAINCK=y
 CONFIG_GPIO=y
 CONFIG_WDT_DISABLE_AT_BOOT=y
+CONFIG_MEMC=y
```
Hi @pdgendt ,

I was looking closely about this and I agreed with you. Let's keep it!

nandojve commented on drivers/memc/CMakeLists.txt:
```diff
@@ -9,6 +9,8 @@ zephyr_linker_sources_ifdef(CONFIG_MEMC_STM32_SDRAM SECTIONS memc_stm32_sdram.ld
 zephyr_library_sources_ifdef(CONFIG_MEMC_MCUX_FLEXSPI memc_mcux_flexspi.c)
 zephyr_library_sources_ifdef(CONFIG_MEMC_MCUX_FLEXSPI_HYPERRAM memc_mcux_flexspi_hyperram.c)
 
+zephyr_library_sources_ifdef(CONFIG_MEMC_SAM_SMC memc_sam_smc.c)
```
Yes, correct! Let's keep as is. My initial thought is that SMC would be same controller that handles SDRAM, which is not the case.

nandojve commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,167 @@
+  Atmel Static Memory Controller (SMC).
+
+  The SMC allows to interface with static-memory mapped external devices such as
+  SRAM, NOR Flash, NAND Flash, SDRAM...
```
Use the full list: SRAM, PSRAM, PROM, EPROM, EEPROM, LCD Module, NOR Flash and NAND Flash.

Drop SDRAM because there is a dedicated controller to handle that.

nandojve commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,167 @@
+  Finally, in order to make the memory available you will need to define new
+  memory device/s in DeviceTree:
+
+  sdram1: sdram@60000000 {
+      compatible = "mmio-sram";
+      device_type = "memory";
+      reg = <0x60000000 DT_SIZE_M(X)>;
+      zephyr,memory-region = "SDRAM1";
```
The EBI allows you to have SDRAM but uses a dedicated controller which is not SMC. Please, fix the example.

nandojve reviewed (changes requested):
> I was looking into this, and there's a test for stm32_sdram which could be reused. However this requires some more effort for the linker sections to be more generic.

If it is possible, better! Project prefers reuse instead add new ones.

> Looking at https://github.com/zephyrproject-rtos/zephyr/pull/42008 and https://github.com/zephyrproject-rtos/zephyr/pull/40422 we could rework that existing test once those PRs are merged.

Your PR will be conditioned to be merged after those.

I was looking this controller and my initial thought was that SMC should handle SDRAM too. After looking datasheets I noted that SAM uses a dedicated controller for SDRAM. So, in this case, we have two distinct drivers and those are independently.

On that research I noted that SAMS7/E7/V7 uses same SMC ip. It is up to you add that nodes like SAM4E. There is a minimal change in the driver that could be added which will allow SAMS7/E7/V7 have 16 bit access instead only 8. If you don't mind add this, you can upstream the whole feature set. This is a plus not a block from my side.

pdgendt commented:
> > Looking at #42008 and #40422 we could rework that existing test once those PRs are merged.
> 
> Your PR will be conditioned to be merged after those.
> 

So we wait a bit, or?

> I was looking this controller and my initial thought was that SMC should handle SDRAM too. After looking datasheets I noted that SAM uses a dedicated controller for SDRAM. So, in this case, we have two distinct drivers and those are independently.

Do you mean on the SMC module as well?

> On that research I noted that SAMS7/E7/V7 uses same SMC ip. It is up to you add that nodes like SAM4E. There is a minimal change in the driver that could be added which will allow SAMS7/E7/V7 have 16 bit access instead only 8. If you don't mind add this, you can upstream the whole feature set. This is a plus not a block from my side.

I currently have limited time to add these additional SoCs, and no hardware to test. The SAM4E looked easy, but I also have no way to test this. I guess if someone needs them, is should be easy to test and upstream.

nandojve commented:
> > Your PR will be conditioned to be merged after those.
> 
> So we wait a bit, or?

Yes, to reuse.

> Do you mean on the SMC module as well?

As far I understood SMC not support SDRAM. SDRAM is handled by a dedicated controller. At end, EBI is a set of one or two controllers (SMC + SDRAM).

nandojve commented:
Hi @pdgendt ,

I would like to add this after #43731.

mbolivar-nordic commented:
Not for v3.1

nandojve commented:
Hi @pdgendt ,

The #43731 is already in and you can update PR with the new pinctrl API support.

pdgendt committed:
boards: arm: sam4s_xplained: Enable SRAM on SMC bus

This commit enables the SRAM controller is66wv51216dbll connected
to the SMC/EBI bus. Half of the memory is made available (512KB)
on chip select 0, the other part is (512KB) on chip select 1.

The SMC timings are taken from the Microchip studio example
SMC_SRAM_EXAMPLE for the sam4s_xplained board.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
> The #43731 is already in and you can update PR with the new pinctrl API support.

Hi @nandojve,

I've updated the PR to use the new pinctrl API, however I don't have the board to verify, is this something you could do?

Thanks

nandojve reviewed (changes requested):
Hi @pdgendt ,

Yes, I can test in the boards. Please, make sure that test/drivers/memc will trigger. Currently it is not.

nandojve commented:
Hi @pdgendt ,

Implementation is OK, tested on sam4s_xplained

```console
*** Booting Zephyr OS build zephyr-v3.1.0-541-ga7b1463ba7aa  ***
Running TESTSUITE smc_sram
===================================================================
START - test_sram1_rw
 PASS - test_sram1_rw in 0.1 seconds
===================================================================
START - test_sram2_rw
 PASS - test_sram2_rw in 0.1 seconds
===================================================================
TESTSUITE smc_sram succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

I was wondering if it is a good idea unify those tests.

pdgendt commented:
> Implementation is OK, tested on sam4s_xplained

@nandojve Nice, thanks for testing!

> I was wondering if it is a good idea unify those tests.

I think that would be best, indeed. But maybe this could be done in a separate PR?

nandojve reviewed (dismissed):
Hi @pdgendt,

> I think that would be best, indeed. But maybe this could be done in a separate PR?

It is OK from my side.

Thank you!

henrikbrixandersen commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,129 @@
+  reg:
+      required: true
+
+  label:
+    required: true
```
Is the label really required?

pdgendt replied:
Not for me, I compared with other similar drivers, and it appears this is added for most.
Will remove.

henrikbrixandersen commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,129 @@
+
+child-binding:
+  description: |
+    Children device nodes are representing devices connected to the EBI/SMC bus.
```
```suggestion
    Child device nodes are representing devices connected to the EBI/SMC bus.
```

henrikbrixandersen commented on tests/drivers/memc/smc_sram/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+  drivers.memc.smc_sram:
+    tags: drivers memc
+    depends_on: memc
+    filter: dt_compat_enabled("atmel,sam-smc")
```
Another option would be to put this under `tests/boards/atmel_sam` or similar - of make it generic for all memc drivers.

pdgendt replied:
This suggestion was already done by @nandojve , but I wanted to this in a separate PR.

henrikbrixandersen replied:
Why not fix it before introducing it?

gmarull replied:
note: `memc` are a set of drivers without an API, so a generic test isn't possible afaik.

henrikbrixandersen replied:
The current test simply tests RW access to a memory region from a device on the memory controller - that sounds like something that could work for at least more than one memc device type.

pdgendt replied:
@henrikbrixandersen ok, to update the test in a follow-up PR?

henrikbrixandersen replied:
I don't see why we should introduce the test with identified issues only to do a follow-up PR. Please address it in this PR.

pdgendt replied:
Done

pdgendt commented:
Rebased and added multi-instance support.

pdgendt committed:
tests: drivers: memc: Rework S(D)RAM test with ztest

* Reworked the stm32 SDRAM test to be more generic
* Added SRAM entries
* Moved to new ztest API

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
@nandojve could you re-test the new implementation with the `xplained` board?
Thanks!

henrikbrixandersen reviewed (dismissed):
Thanks :)

nandojve commented:
> @nandojve could you re-test the new implementation with the `xplained` board? Thanks!

Hi @pdgendt ,
I`m without HW this week. I can run tests after July 20th when I return from a trip.
Overall LGTM!

carlescufi commented:
@nandojve can you take a look?

pdgendt commented:
> Hi @pdgendt , I`m without HW this week. I can run tests after July 20th when I return from a trip. Overall LGTM!

Hey @nandojve , friendly reminder to revisit this :)

galak commented on dts/arm/atmel/sam4e.dtsi:
```diff
@@ -253,6 +253,16 @@
+			reg = <0x40060000 0x200>;
+			peripheral-id = <8>;
+			status = "disabled";
+			label = "SMC";
```
remove label prop

galak commented on dts/arm/atmel/sam4s.dtsi:
```diff
@@ -211,6 +211,16 @@
+			reg = <0x400e0000 0x200>;
+			peripheral-id = <10>;
+			status = "disabled";
+			label = "SMC";
```
remove label prop

galak commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,126 @@
+
+  properties:
+    reg:
+      type: int
```
Does `reg` here represent a chip select or something?

pdgendt replied:
Yes, it's a chip select.

galak replied:
maybe add a `description` field to `reg` and mention that there.

galak commented on drivers/memc/Kconfig.sam:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2022 Basalte bv
+# SPDX-License-Identifier: Apache-2.0
```
Update Kconfig to new pattern based on autogenerated <BUILD_DIR>/Kconfig/Kconfig.dts.  See https://github.com/zephyrproject-rtos/zephyr/pull/48169 for example of the change.

galak reviewed (dismissed):
Some minor Kconfig & dts changes.

galak commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,129 @@
+      default: [3, 3]
+      description: |
+        SMC cycle timing configuration. Refer to the datasheet for more
+        information. Expected fields, in order, are,
```
remove white space.  Also, not sure how `fields` NWE/NRD and an array quite make sense together.

pdgendt replied:
I used the array type to create a compact view instead of an entry for each item, and the values are set into the same HW register.

galak replied:
The array is fine, I just find the wording confusing.  not sure if `[NWE NRD]` in the description would make it a little clearer.

galak commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,129 @@
+    atmel,smc-write-mode:
+      type: string
+      required: false
+      default: "nwe"
```
Do we need `default:` for all these properties?  What are these default values based on?

pdgendt replied:
The default values are based on the datasheet, the register data after reset.

galak reviewed (commented):
some comments on the binding.

nandojve commented on tests/drivers/memc/ram/testcase.yaml:
```diff
@@ -3,3 +3,7 @@ tests:
     tags: drivers memc
     depends_on: memc
     filter: dt_compat_enabled("st,stm32-fmc-sdram")
+  drivers.memc.smc_sram:
+    tags: drivers memc
+    depends_on: memc
+    filter: dt_compat_enabled("atmel,sam-smc")
```
twister is skipping `sam4s_xplained` board. It should be possible build locally with, I think:
`./scripts/twister -p sam4s_xplained -t tests/drivers/memc/ram`

pdgendt replied:
Shouldn't this work? I'm not sure how to fix/improve this.
I thought about dropping the `dt_compat_enabled` filter entirely and check the nodelabels instead:

```yaml
tests:
  drivers.memc.ram:
    tags: drivers memc
    depends_on: memc
    filter: dt_nodelabel_enabled("sdram1") or dt_nodelabel_enabled("sdram2") or
            dt_nodelabel_enabled("sram1") or dt_nodelabel_enabled("sram2")

```

Any suggestions?

nandojve replied:
No clue about it. Please, take a look at Discord -> Infrastructure channel.

pdgendt replied:
@nandojve I've added `sam4s_xplained` as an integration platform, CI should build it now.

nandojve reviewed (changes requested):
Hi @pdgendt ,

I run on board and all tests passed, see below:
```
*** Booting Zephyr OS build zephyr-v3.1.0-2551-g2faab4602eff  ***
Running TESTSUITE test_ram
===================================================================
START - test_sdram1
 SKIP - test_sdram1 in 0.1 seconds
===================================================================
START - test_sdram2
 SKIP - test_sdram2 in 0.1 seconds
===================================================================
START - test_sram1
 PASS - test_sram1 in 0.1 seconds
===================================================================
START - test_sram2
 PASS - test_sram2 in 0.1 seconds
===================================================================
TESTSUITE test_ram succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

Please, check why twister is skipping the test.

mbolivar-nordic commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,116 @@
+        - "ncs"
+        - "nrd"
+
+    atmel,smc-setup-timing:
+      type: array
+      required: false
+      default: [1, 1, 1, 1]
+      description: |
+        SMC setup timing configuration. [NWE, NCS_WR, NRD, NCS_RD]
+        Refer to the datasheet for more information.
+
+    atmel,smc-pulse-timing:
+      type: array
+      required: false
+      default: [1, 1, 1, 1]
+      description: |
+        SMC pulse timing configuration. [NWE, NCS_WR, NRD, NCS_RD]
+        Refer to the datasheet for more information.
+
+    atmel,smc-cycle-timing:
+      type: array
+      required: false
+      default: [3, 3]
+      description: |
+        SMC cycle timing configuration. [NWE, NRD]
+        Refer to the datasheet for more information.
```
All of these defaults must be justified in the descriptions.  https://docs.zephyrproject.org/latest/build/dts/bindings.html#rules-for-default-values

pdgendt replied:
@mbolivar-nordic I've updated the descriptions with information from the datasheet

mbolivar-nordic commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,130 @@
+    required: true
+    const: 0
+
+  peripheral-id:
```
Is this actually going into a register somewhere or are you just using it to work around hal limitations? If the former, it's OK, but if the latter, then @galak has a longstanding prohibition against such properties.

pdgendt replied:
The `peripheral-id`? This is how all the `sam` SoCs identify a HW block, and isn't something new. I don't want to turn this PR into something bigger.

nandojve replied:
Peripheral-is represents a bit value used to enable clock on the peripheral. For each SoC model this value change the current implementation it is the easy solution to allow share same driver between SoCs and allow DT_INST_FOREACH_STATUS_OKAY for instantiate. I think this property could be removed in the future after implement a clock driver and clock tree.

mbolivar-nordic replied:
> I think this property could be removed in the future after implement a clock driver and clock tree.

@nandojve based on our offline discussion, it sounds like you're agreeing to take this on so that these properties can be removed treewide for SAM SoCs. Given maintainer commitment to improve this, I am OK with this for now.

mbolivar-nordic commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,130 @@
+      required: false
+      default: [1, 1, 1, 1]
+      description: |
+        SMC setup timing configuration in cycles. [NWE, NCS_WR, NRD, NCS_RD]
+        1 cycle is the default for devices that perform actions on rising
+        edges. Null delay setups are permitted, for consecutive reads/writes.
```
This is kind of confusing still. It says 1 is the default but the default is actually an array. I guess the description is meant to say the first element is NWE setup timing, the second is NCS_WR, etc.? I personally find that a bit confusing if so.

My bigger question here though is why are you assuming devices that perform actions on rising edges? Would a different configuration only appear under very unusual conditions? If it's common for devices not to perform actions on rising edges, then these should be made required instead.

pdgendt replied:
I was trying to translate the datasheet a bit into the description and I thought it would be safer to have a minimal setup timing as the default for all read/write modes.
<img width="563" alt="image" src="https://user-images.githubusercontent.com/3170234/181699730-8e09f6ee-d145-4ef0-9559-bc2dd7df9664.png">

Maybe our atmel/michrochip maintainer @nandojve can suggest how the (timing) properties should be set?

Sorry this has taken already so many iterations, but I agree it has to be clear for everyone.

nandojve replied:
Hi @pdgendt ,

The devicetree properties must be well documented. A maintainer from devicetree should read the documentation and have a very clear picture about it without inspect code or even datasheet. Otherwise no one can provide fixes in some situations and it is even more important when the property have a **default** value. I think both @mbolivar-nordic and @galak are arguing in that direction to make sure that properties are well defined and well documented.

In regards of timing parameters, I'm OK to have it as a array. I think what @mbolivar-nordic is asking to you go in the direction of have 100% sure that a  `default` is fine for this driver since the documented is not 100% clear. The default should be used only when you have a generic case that can full fill majority of scenarios otherwise `required` field should be added and set as `true`. This imposes that each board must explicit define the values since no `default` property will be available. The other think that I noted is that for someone that is not aware about SoC it is difficult to understand the default values.

If you look at my current PR #48377 you will noted that I'm in the same situation. The amount of information that I added at first push wasn't enough to make it clear. A second version was send with substantial more documentation and examples. This kind of situation happen because, in your case, the driver is one of the first and maintainers will look in deep to make sure it is clear how properties were define. Remember that people inspect code to pick ideas for drivers and this maybe be a reference for others.

Questions that may help to evaluate the properties:

* What happen if user selects another SDRAM manufacturer? Are the parameters same?
* What happen if user selects a slower speed memory? The current default configuration is set up for the slowest case?
* The current default values may work for all memories or at least for the majority of that?
* If I set up SAM4E-XPRO board will be possible to use same SDRAM parameters from SAM4S-Xplained (since memories are different)?

Sugestions:

 * Increase the amount of documentation
 * Add the sam4s_xplained configuration as an example in the devicetree documentation
 * Explain how to configure each parameter and why you select that value since it have a relationship with memory used and the SoC controller.

pdgendt replied:
Thanks @nandojve for the elaborate response.

Based on the feedback, it is clear that default values do not make sense, due to the amount of different external controllers possible.
All properties are now required without defaults.

Is the `sam4s_xplained` example configuration in the devicetree documentation sufficient?

mbolivar-nordic reviewed (changes requested):
Still keeping a -1 on for now because it's not clear to me if this is a correct use of default.

nandojve reviewed (changes requested):
Hi @pdgendt ,

I think it is better but still missing important information. I tried highlight few points to you.

nandojve commented on dts/bindings/memory-controllers/atmel,sam-smc.yaml:
```diff
@@ -0,0 +1,140 @@
+      type: array
+      required: true
+      description: |
+        SMC pulse timing configurations in cycles for the signals NWE, NCS_WR,
+        NRD and NCS_RD, in that order. Null pulses are not permitted and lead
+        to unpredictable behavior.
+        Encoded in 7 bits where the highest bit adds an offset of 256.
+        Effective value for each element: 256 x pulse[6] + pulse[5:0]
```
This value is used to effectivelly read/write at memory region (pulse phase).
The pulse value is an array of the signals NWE, NCS_WR, NRD and NCS_RD were
each value is configured in terms of MCK cycles and value of 0 is forbiden.
Each value is encoded in 7 bits where the highest bit adds an offset of 256
cycles. The effective value for each element is: 256 x setup[6] + setup[5:0]

nandojve reviewed (changes requested):
Hi @pdgendt ,

It still not clear. Please, see my suggestions and adapt as you wish.

pdgendt commented:
Thanks @nandojve, for all your efforts to help out!

I updated the suggestions you made, but adjusted the example timings to the datasheet specifications.
A minimum setup time of 5ns and puls time of 45ns where swapped in your description?

nandojve commented:
Hi @pdgendt ,

I`m OK with changes. Let`s ask to @mbolivar-nordic to check if fits.

BTW, we need fix flash stuff

nandojve commented:
> BTW, we need fix flash stuff

Fixed by #48762

nandojve commented:
Hi @mbolivar-nordic ,

No more CI erros and @pdgendt made few changes. I was wondering if you can revisit this PR.

mbolivar-nordic reviewed (approved):
@pdgendt thanks for the big improvements. I'm glad to see the properties go to `required`. The `default` feature has its uses, but it is easy to abuse and I'm glad for your and @nandojve figuring out a better way forward.

talih0 commented:
@pdgendt, @nandojve, I'm working on a similar driver for Infineon XMC4xxx.
A problem I've come across is that it's not possible to use zephyr_code_relocate() to move data, bss to external RAM with these type of drivers.

The driver set up happens with POST_KERNEL priority, but 
bss_zeroing_relocation(); and  data_copy_xip_relocation() are called very early at boot up. This driver should have the same problem. 

Did you come across something like this?

pdgendt commented:
> @pdgendt, @nandojve, I'm working on a similar driver for Infineon XMC4xxx. A problem I've come across is that it's not possible to use zephyr_code_relocate() to move data, bss to external RAM with these type of drivers.
> 
> The driver set up happens with POST_KERNEL priority, but bss_zeroing_relocation(); and data_copy_xip_relocation() are called very early at boot up. This driver should have the same problem.
> 
> Did you come across something like this?

Hey @talih0, I did not need relocating to external RAM, so I can't really share my experience there. If this is something you want to be looked at, I would suggest creating a new issue. Comments on merged PRs are easily overlooked :)

talih0 commented:
thanks @pdgendt for following up, I'll create an issue.
